### PR TITLE
Adds monitoring of services when opening ui to eventually fail fast

### LINKED ISSUE #147: Fail fast during sourced initialization if a component breaks
If a component fails during **source{d} CE** initialization (before opening the browser, during the spinner spins), the process should fail

### PR DESCRIPTION
**Pull request by se7entyse7en**
This closes #147.

When the spinner is running and the user is waiting for the ui to open, a monitoring of the state of the containers is performed in order to stop if an unexpected state occurs.

I also ran integration tests on windows.

### CODE REVIEW

**Comment by se7entyse7en** [comment]
Regarding this PR, I've added latest docker-sdk as dependency, but it required me to also update docker. Do you think it's better to use an older version?

File: cmd/sourced/cmd/web.go [hunk 1]
```diff
@@ -75,11 +55,117 @@ func openUI() error {
...
+		}
+
+		containerIds := strings.Split(
+			normalizeNewLine(strings.TrimSpace(stdout.String())),
+			"\n",
+		)
+
+		for {
+			for _, contID := range containerIds {
+				cont, err := docker.InfoWithClient(client, contID)
```
Comment (smacker) [inline cmd/sourced/cmd/web.go]: why do you need calling docker directly if `docker-compose ps` which you run above already returns you all the statuses?

File: cmd/sourced/docker/docker.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	var info types.Info
+	// Get information from running daemon to check whether is running
+	// docker toolbox
+	info, err = c.Info(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	if strings.Contains(strings.ToLower(info.OperatingSystem), "boot2docker") {
+		return nil, fmt.Errorf("Docker Toolbox is not supported")
```
Comment (smacker) [inline cmd/sourced/docker/docker.go]: can you remind me why we don't support Docker Toolbox?
If we don't interact with docker directly but only through `docker-compose` it should work, no?

File: cmd/sourced/cmd/web.go (hunk 1 above)
Comment (se7entyse7en) [reply cmd/sourced/cmd/web.go]: the reason is that I'd have to parse the stdout and I found it not much reliable, that's why I decided to use the client instead.

File: cmd/sourced/docker/docker.go (hunk 2 above)
Comment (se7entyse7en) [reply cmd/sourced/docker/docker.go]: mhmm... Actually, I copy-pasted the code from engine as you without thinking at the reason. But that's definitely something to rethink about. I don't remember very well the reason, AFAIR it has something to do with Windows support, but I have to recheck it.
If this limitation doesn't apply by using only `docker-compose`, then for sure it's better to avoid using the client and parse the stdout.

Comment (se7entyse7en) [reply cmd/sourced/docker/docker.go]: @carlosms @dpordomingo do you perhaps the remember the reason? I don't remember whether we _really_ do not support docker toolbox or whether we just decided to not support it.

**Comment by se7entyse7en** [comment]
As [discussed in TFM](https://github.com/src-d/minutes/pull/681/files#diff-d685f063c7833b25d1843b8d6db0885aR60), we're going to remove Go SDK client in favor of grepping from stdout.

**Comment by se7entyse7en** [comment]
@carlosms @smacker I updated and force pushed as the implementation completely changed.

File: cmd/sourced/cmd/web.go [hunk 3]
```diff
@@ -32,54 +33,177 @@ func init() {
...
+// an unrecoverable error occurs.
+// The monitoring is performed by running `docker-compose ps <service>` for each
+// service returned by `docker-compose config --services`, and by grepping the
+// state from the stdout using a regex.
+// Getting the state of all the containers in a single pass by running `docker-compose ps`
+// and by using a multi-line regex to extract both service name and state is not reliable.
+// The reason is that the prefix of a container can be very long, especially for local
+// initialization, due to the value that we set for `COMPOSE_PROJECT_NAME` env var, and
+// docker-compose may split the name into multiple lines.
+// E.g.:
```
Comment (se7entyse7en) [inline cmd/sourced/cmd/web.go:94]: Do you think that the example is too much?

**Comment by se7entyse7en** [comment]
I've manually tested this by running `sourced init local <path>` on a shell, and by running `docker kill <gitbase container id>` in another.

File: cmd/sourced/cmd/web.go [hunk 4]
```diff
@@ -32,54 +33,177 @@ func init() {
...
+		for _, service := range services {
+			go runMonitorService(service, ch)
+		}
+	}()
 }
 
 // OpenUI opens the browser with the UI.
 func OpenUI(timeout time.Duration) error {
+	var stdout bytes.Buffer
+	failFast, err := checkFailFast(&stdout)
```
Comment (smacker) [inline cmd/sourced/cmd/web.go:159]: it took me a while to figure what is going on here. Do you think it would be easier to skip only error "container not available" instead? And fail for all other errors. Also, we might want to have the same check inside `waitForContainer`. Currently if, for example, docker server crashed we will ignore the error.

File: cmd/sourced/cmd/web.go [hunk 5]
```diff
@@ -32,54 +33,177 @@ func init() {
...
 			break
 		}
 
 		time.Sleep(1 * time.Second)
 	}
+}
 
-	if err := browser.OpenURL(url); err != nil {
-		return errors.Wrap(err, "could not open the browser")
+var newLineFormatter = regexp.MustCompile(`(\r\n|\r|\n)`)
```
Comment (smacker) [inline cmd/sourced/cmd/web.go]: why do you need to do it manually? `strings.TrimSpace` already remove all space characters: https://golang.org/pkg/unicode/#IsSpace

File: cmd/sourced/cmd/web.go (hunk 3 above)
Comment (smacker) [reply cmd/sourced/cmd/web.go:94]: yes :) just 1 line is enough to understand.

File: cmd/sourced/cmd/web.go [hunk 6]
```diff
@@ -32,54 +33,177 @@ func init() {
...
+				}
+
+			}
+
+			time.Sleep(1 * time.Second)
+		}
 	}
 
-	return nil
+	go func() {
```
Comment (smacker) [inline cmd/sourced/cmd/web.go]: let's move `go` out of this function, it will be easier to understand and would allow to use the function in sync mode if needed

File: cmd/sourced/cmd/web.go [hunk 7]
```diff
@@ -32,54 +33,177 @@ func init() {
...
+			os.Stdin, &servicesBuf, nil, "config", "--services"); err != nil {
+			ch <- errors.Wrap(err, "cannot get list of services")
+			return
+		}
+
+		services := strings.Split(normalizeNewLine(
+			strings.TrimSpace(servicesBuf.String())), "\n")
+
+		for _, service := range services {
+			go runMonitorService(service, ch)
```
Comment (smacker) [inline cmd/sourced/cmd/web.go]: what is the advantage of using goroutines here?

File: cmd/sourced/cmd/web.go [hunk 8]
```diff
@@ -32,54 +33,177 @@ func init() {
...
 		}
 
 		time.Sleep(1 * time.Second)
 	}
+}
 
-	if err := browser.OpenURL(url); err != nil {
-		return errors.Wrap(err, "could not open the browser")
+var newLineFormatter = regexp.MustCompile(`(\r\n|\r|\n)`)
+var stateExtractor = regexp.MustCompile(`(?m)^srcd-[\w\d]+.*(Up|Exit (\w+))`)
```
Comment (smacker) [inline cmd/sourced/cmd/web.go]: should it be `(Exit (\d+)`?

File: cmd/sourced/cmd/web.go (hunk 6 above)
Comment (smacker) [reply cmd/sourced/cmd/web.go]: it's mentioned [here](https://github.com/golang/go/wiki/CodeReviewComments#synchronous-functions) but not very clear. There was a better explanation somewhere (most probably in go blog or somewhere else, I can't find it right now). The idea is it's better to keep functions blocking and if you need to run them in non-blocking way the caller can do it which is much easier to reason about.

File: cmd/sourced/cmd/web.go (hunk 5 above)
Comment (se7entyse7en) [reply cmd/sourced/cmd/web.go]: I've just tried it and yes no need to do it manually. I thought that it would have replace `\r\n` with two new lines.

File: cmd/sourced/cmd/web.go (hunk 6 above)
Comment (se7entyse7en) [reply cmd/sourced/cmd/web.go]: Seems fair, thanks for link!

File: cmd/sourced/cmd/web.go (hunk 8 above)
Comment (se7entyse7en) [reply cmd/sourced/cmd/web.go]: Ops, nice catch! And also the `\d` is redundant in the first part, in the end it will be as follows:
```
(?m)^srcd-\w+.*(Up|Exit (\d+))
```

File: cmd/sourced/cmd/web.go (hunk 7 above)
Comment (se7entyse7en) [reply cmd/sourced/cmd/web.go]: The only reason is to be more reactive to state changes in containers. The alternative that I thought is to remove the `while true` and the `sleep` from the `runMonitorService` function and to write this for loop as:
```
for {
    for _, service := range services {
        runMonitorService(service, ch)
    }
}
```

The problem is that the call to `docker-compose ps <service>` takes an amount of time near to ~1 second (at least in my setup). So with the non-gorountine solution, the container of a service is checked once every ~`n`-seconds where `n` is the number of services. With the gorountine solution empirically each service each checked much faster. With the `sleep` of 1 second as now, every service is checked every ~4 seconds. It's to be noted that this is based only on empirical observations and that I'm pretty sure also for the gorountine solution the amount of time is a function of `n` too.
Long story short, with our current number of services docker answers faster to all our requests if they're made concurrently instead of serially. This may change if in future we add more services.

Comment (se7entyse7en) [reply cmd/sourced/cmd/web.go]: So if we're happy with a less reactive solution in favor of having easier code, I'm also ok with that. I don't think that a user would get mad for waiting a couple of seconds more.

File: cmd/sourced/cmd/web.go (hunk 4 above)
Comment (se7entyse7en) [reply cmd/sourced/cmd/web.go:159]: > it took me a while to figure what is going on here. Do you think it would be easier to skip only error "container not available" instead? And fail for all other errors.

So you mean that we could avoid checking for those 3 types of errors? It makes sense to me. What do you think @dpordomingo? I'm asking you as you recently refactored this part regarding error handling AFAIR. I don't know whether there are other errors that we want to skip in addition to "container not available".

> Also, we might want to have the same check inside `waitForContainer`. Currently if, for example, docker server crashed we will ignore the error.

In this case, the monitor should correctly raise an error and exit.

File: cmd/sourced/cmd/web.go (hunk 3 above)
Comment (se7entyse7en) [reply cmd/sourced/cmd/web.go:94]: Done [here](https://github.com/src-d/sourced-ce/pull/186/commits/f98097944933f458fe2ceed62ce80173cd6b2c2c).

File: cmd/sourced/cmd/web.go (hunk 5 above)
Comment (se7entyse7en) [reply cmd/sourced/cmd/web.go]: Done [here](https://github.com/src-d/sourced-ce/pull/186/commits/fd92593689fc5ddf7f8abcc58086ef2659634042).

File: cmd/sourced/cmd/web.go (hunk 6 above)
Comment (se7entyse7en) [reply cmd/sourced/cmd/web.go]: Done [here](https://github.com/src-d/sourced-ce/pull/186/commits/789f447283e70ac23ec272e92d8c5064dfa635f6).

File: cmd/sourced/cmd/web.go (hunk 8 above)
Comment (se7entyse7en) [reply cmd/sourced/cmd/web.go]: Done [here](https://github.com/src-d/sourced-ce/pull/186/commits/2a9ce5425fe437c68c458a0beec0818d61c4d387).

File: cmd/sourced/cmd/web.go (hunk 4 above)
Comment (smacker) [reply cmd/sourced/cmd/web.go:159]: Currently, we are checking for 3 errors that we know how to display better. But the function can fail with many other errors we don't know about. They are just ignored now.

I propose to invert the logic. Check only for errors we want to skip and return everything else "fast" (including this 3 errors).

Maybe it's not necessary to do in this PR.

Comment (smacker) [reply cmd/sourced/cmd/web.go:159]: > In this case, the monitor should correctly raise an error and exit.

that is true. But it's a side effect...

File: cmd/sourced/cmd/web.go (hunk 7 above)
Comment (smacker) [reply cmd/sourced/cmd/web.go]: I don't have a strong opinion about one or another approach.
If I was writing it from scratch I would choose simpler code over 2 seconds delay of error reporting (taking into account that the user sees the loader at this time). But if you prefer to keep it as it is now, it's fine.

File: cmd/sourced/cmd/web.go (hunk 4 above)
Comment (se7entyse7en) [reply cmd/sourced/cmd/web.go:159]: > I propose to invert the logic. Check only for errors we want to skip and return everything else "fast" (including this 3 errors).

This makes sense. 👍 

> that is true. But it's a side effect...

Yeah, it's a side effect. The "problem" is that even if we add that check in the `waitForContainer` function, probably will be still the monitor to catch that error as it runs more frequently. But having that check even in `waitForContainer` will surely increase the understandability of the flow, so I'm ok adding it.

File: cmd/sourced/cmd/web.go (hunk 7 above)
Comment (se7entyse7en) [reply cmd/sourced/cmd/web.go]: I'm ok with both really 😄. Let's see also what @dpordomingo and @carlosms say then. 👍

**Review by smacker — APPROVED** [review]

**Comment by carlosms** [comment]
LGTM, but let's add an entry to the changelog please.

**Comment by se7entyse7en** [comment]
> LGTM, but let's add an entry to the changelog please.

I completely missed this 😞. I'll do it ASAP.

**Comment by se7entyse7en** [comment]
Rebased and added changelog entry [here](https://github.com/src-d/sourced-ce/pull/186/commits/d0214bf7e11ca7a3cfae821685d8b17772429e63).

File: CHANGELOG.md [hunk 9]
```diff
@@ -3,8 +3,9 @@
 ## [Unreleased]
 <details>
   <summary>
-    Changes that have landed in master but are not yet released.
-    Click to see more.
+
+- Add a monitoring of containers state while waiting for the web UI to open during initialization.
```
Comment (carlosms) [inline CHANGELOG.md]: The previous contents did not use the changelog categories because it was the initial release, but I think we should start from now on.
```suggestion
### Added

- Add a monitoring of containers state while waiting for the web UI to open during initialization ([#147](https://github.com/src-d/sourced-ce/issues/147)).
```

**Comment by se7entyse7en** [comment]
@carlosms force pushed [here](https://github.com/src-d/sourced-ce/pull/186/commits/68bb53251212c65281d09ba44bd5a28837905651).

**Review by carlosms — APPROVED** [review]
